# Standard Analysis

Transcription Start Sites (TSSs) represent the first base to be transcribed by RNA Polymerase.
Most genes tend to have a heterogenous collection of TSSs as opposed to a single position.
This is interesting because choice of TSS can affect gene isoform, stability, and translational efficiency.
Furthermore, TSS choice has been implicated in develoment, homseostasis, and disease.

tsrexploreR has a series of analysis and plotting functions to allow deep exploration of TSSs.

## Preparing Data

This example will use a set of *S. cerevisiae* TSSs collected using the STRIPE-seq method.
There are many ways to import TSSs into tsrexplorer.
This example data uses a named list of GRanges as imput into the function that creates the tsrexplorer object.

```
library("tsrexplorer")
library("magrittr")

TSSs <- system.file("extdata", "S288C_TSSs.RDS", package = "tsrexplorer")
TSSs <- readRDS(TSSs)

# Keep only the 3 WT samples for now.
TSSs <- names(TSSs) %>%
	stringr::str_detect("WT") %>%
	purrr::keep(TSSs, .)

exp <- tsr_explorer(TSSs)
```

## Initial TSS Processing

After the TSSs are loaded into the tsrexplorer object,
there are a few processing steps to go through to get the data ready for analysis.

### Format TSSs

The first step is to convert the TSSs into a table that will facilitate downstream analysis.

```
exp <- format_counts(exp, data_type = "tss")
```

### Normalize TSSs

The next step is to Counts Per Million (CPM) normalize the TSSs.
This step is optional, and if the counts you inputed were normalized already this step can safely be skipped.

```
exp <- cpm_normalize(exp, data_type = "tss")
```

### TSS Annotation

After formatting the counts and optionally CPM normalizing them, the TSSs will be annotated relative to known features.
This function takes either the path and file name of a 'GTF' or 'GFF' file, or a 'TxDb' package from bioconductor.
The example below uses a 'GTF' file from Ensembl (R64-1-1 Release 99), and will annotate each TSS to the closest transcript.

```
annotation <- system.file("extdata", "S288C_Annotation.gtf", package = "tsrexplorer")

exp <- annotate_features(
        exp, annotation_data = annotation,
        data_type = "tss", feature_type = "transcript"
)
```

### Naive Threshold Exploration

In most TSS mapping methods there are low count TSSs that are difficult to distinguish between signal and noise.
A good amount of this background can be removed by requiring a certain number of reads per TSS.
If a genome annotation is available, a thresholding plot may help in picking this read threshold.

```
threshold_data <- explore_thresholds(exp, max_threshold = 25)

p <- plot_threshold_exploration(threshold_data, ncol = 3, point_size = 0.5) +
	ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tss_thresholding.png", plot = p, device = "png", type = "cairo", height = 1.25, width = 5)
```

![tss_thresholds](../inst/images/tss_thresholding.png)

Looking at the plot, a threshold of three or more reads is a sensible choice.
There is a precipitous drop in detected features for a comparatively low gain in
promoter proximal fraction of TSSs at higher thresholds.

## TSS Correlation

Finding the correlation between samples can be informative of both sequencing efficacy,
and to also get a cursory understanding about the differences between different samples.

### TMM Normalization

For correlation analysis, tsrexploreR uses 'TMM' normalization from the edgeR library.
This method was designed to make comparison **between** samples more efficacious.
In the example below, TSSs are retained only if at least one sample has a count of 3 or more reads.

```
exp <- count_matrix(exp, data_type = "tss")
exp <- tmm_normalize(exp, data_type = "tss", threshold = 3, n_samples = 1)
```

### Correlation Matrix Plots

After 'TMM' normalizing between the samples, various correlation plots can be generated by tsrexplorer.
An example is shown here, in which half the plot is a scatter plot, and the other half a heatmap.

```
p <- plot_correlation(exp, data_type = "tss", font_size = 2, pt_size = 0.4) +
        ggplot2::theme_bw() +
        ggplot2::theme(text = element_text(size = 3), panel.grid = element_blank())

ggsave("tss_correlation.png", plot = p, device = "png", type = "cairo", height = 2, width = 2)
```

![tss_corr_plot](../inst/images/tss_correlation.png)

## TSS Genomic Distribution

As part of the TSS processing steps, TSSs were annotated relative to known features.
This information can be used to explore the distribution of TSS throughout the genome,
as well as information on detected features.

### Genomic Distribution Plot

A stacked bar plot can be generated to showcase the fractional distribution of TSSs relative to features.

```
tss_distribution <- genomic_distribution(exp, data_type = "tss", threshold = 3)

p <- plot_genomic_distribution(tss_distribution) +
	ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tss_genomic_distribution.png", plot = p, device = "png", type = "cairo", height = 1, width = 2.5)
```

![tss_genomic_distribution](../inst/images/tss_genomic_distribution.png)

### Feature Detection Plot

The number of genes, and fraction of genes with a promoter proximal TSSs can be made into a stacked bar plot.

```
features <- detect_features(exp, data_type = "tss", threshold = 3)

p <- plot_detected_features(features) +
	ggplot2::theme(text = element_text(size = 3), legend.key.size = unit(0.3, "cm"))

ggsave("tss_feature_plot.png", plot = p, device = "png", type = "cairo", height = 1, width = 1.75)
```

![tss_feature_plot](../inst/images/tss_feature_plot.png)

### Average Plots

Another useful plot type for TSSs are average plots centered around annotated TSSs.
The current yeast genome annotation does not contain any information on 5' or 3' UTRs,
thus the average plot is centered on annotated start codons.
Because of this one would expect the average plot to be slightly upstream of the start codon center point.
Most other organisms have UTRs in their genome annotation,
with the UTR length being the furthest TSS detected from the start codon.
This would then result in average plots that are expected to be centered and slightly downstream from the annotated TSS center.

```
p <- plot_average(exp, data_type = "tss", threshold = 3, ncol = 3) +
	ggplot2::theme(text = element_text(size = 4))

ggsave("tss_average_plot.png", plot = p, device = "png", type = "cairo", height = 1, width = 2)
```

![tss_average_plot](../inst/images/tss_average_plot.png)

### Heatmaps

While an average plot may give a general overview of TSSs relative to annotated start codons or TSSs,
it may sometimes be appropriate to generate a heatmap with TSS positions for all features displayed.
Due to TSSs being single points and sometimes sparsely distributed, it may be hard to them on a heatmap.
This option is provided regardless.

```
count_matrix <- tss_heatmap_matrix(exp, threshold = 3, upstream = 250, downstream = 250)

p <- plot_heatmap(count_matrix, ncol = 3, background_color = "white") +
	ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tss_heatmap.png", plot = p, device = "png", type = "cairo", height = 2, width = 4)
```

![tss_heatmap](../inst/images/tss_heatmap.png)

## Sequence Analysis

TSSs tend to occur in certain sequence contexts, and this context can vary between species.
Knowing this bias can give mechanistic and biologically relevant information on promoter structure.

### TSS Sequence Logo

Generating sequence logos around TSSs is a good preliminary step to better understand the sequence context of TSSs.
For example, in *S. cerevisiae* it has been previously published that there is a pyrimidine-purine bias in the -1 and +1 positions respectively.
Furthermore, stronger TSSs tend to have a well position adenine in the -8 position, the loss of which diminishes promoter strength.

First, the sequences centered around TSSs will be retrieved using a 'FASTA' genome assembly or'BSgenome' object.
This example uses he Ensembl R64-1-1 Release 99 assembly FASTA.

```
assembly <- system.file("extdata", "S288C_Assembly.fasta", package = "tsrexplorer")

seqs <- tss_sequences(exp, genome_assembly = assembly, threshold = 3)
```

After the sequences are retrieved, the sequence logos can be generated.

```
p <- plot_sequence_logo(seqs, ncol = 3)

png("tss_seq_logo.png", units = "in", res = 300, height = 1, width = 6, type = "cairo")
p
dev.off()
```

![tss_sequence_logo](../inst/images/tss_seq_logo.png)

### Sequence Color Map

A sequence logo "averages" the bases when displaying data, but it can be useful for a more raw visualization.
Sequence color maps will assign a color to each base, and then display the corresponding colors centered around TSSs.
The prevalence of colors in certain positions can give further evidence towards putative sequence contexts.
The same genome assembly and retrieved sequences that were used to make the sequence logos above will be used here.

```
p <- plot_sequence_colormap(seqs, ncol = 3) +
	ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tss_seq_colormap.png", plot = p, device = "png", type = "cairo", height = 1.5, width = 3)
```

![tss_sequence_colormap](../inst/images/tss_seq_colormap.png)

### Dinucleotide Frequency

As previously discused, in most organisms there seems to be a strong pyrimidine-purine bias in the -1 and +1 positions respectively.
A function to explore the -1/+1 dinucleotide frequencies is included to further explore this.

```
assembly <- system.file("extdata", "S288C_Assembly.fasta", package = "tsrexplorer")

frequencies <- dinucleotide_frequencies(exp, genome_assembly = assembly, threshold = 3)

p <- plot_dinucleotide_frequencies(frequencies, ncol = 3) +
	ggplot2::theme(text = element_text(size = 6))

ggsave("tss_dinucleotide_frequencies.png", plot = p, device = "png", type = "cairo", height = 2, width = 5)
```

![tss_dinucleotide_frequencies](../inst/images/tss_dinucleotide_frequencies.png)

## Clustering TSSs

TSSs are often clustered together into Transcription Start Regions (TSRs).
These clusters tend to capture the nature of the trancsription landscape better than looking at individual TSSs.
Furthermore, cluster shape features have been shown to correlate with various functional gene control classes.

### Distance Clustering

TSSexploereR has a distance clustering algorithm built in.
This approach clusters TSSs that pass a read threshold and are within a certain distance.
If desired, TSRs dervived from more advanced clustering methods, such as Paraclu and RECLU from CAGEr, can be imported directly.

```
exp <- tss_clustering(exp, threshold = 3, max_distance = 25)
```

### Associating TSSs with TSRs

TSSexploreR provides great flexibility when working with TSSs and TSRs.
After TSRs are called or imported, TSSs can be assigned to TSRs they overlap with.
Here TSSs will be associated with the TSRs that were derived from them.
However, the TSSs could easilly be associated with merged or consensus TSRs.

```
exp <- associate_with_tsr(
	exp, use_sample_sheet = FALSE,
	sample_list <- list(
		"S288C_WT_1" = "S288C_WT_1",
		"S288C_WT_2" = "S288C_WT_2",
		"S288C_WT_3" = "S288C_WT_3"
	)
)
```

## Initial TSR Processing

Now that the TSSs have been clustered and associated with TSRs, the TSRs can undergo initial processing.
This involves annotating the TSRs, calculating various metrics of TSR shape and strength,
and optional normalization of the TSR scores.

### TSR Metrics

After TSSs have been associated with TSRs, various metrics describing TSR shape and strength can be computed.
First, the dominant TSS within a TSR will be annotated in the dataset.
Then, various shape features such as IQR, shape index, and balance will be calculated.

```
exp <- mark_dominant(exp, data_type = "tss", threshold = 3)
exp <- tsr_metrics(exp)
```

### Normalize TSRs

The TSRs can be optionally CPM normalized.

```
exp <- cpm_normalize(exp, data_type = "tsr")
```

### Annotate TSRs

It is recommended to also annotate the TSRs if a genome annotation is available.

```
annotation <- system.file("extdata", "S288C_Annotation.gtf", package = "tsrexplorer")

exp <- annotate_features(
        exp, annotation_data = annotation,
        data_type = "tsr", feature_type = "transcript"
)
```

## TSR Correlation

Smiilar to looking at the correlation of TSSs, looking at TSR correlation before further analysis is good practice.

### TMM Normalization

TMM normalization is also used for comparing TSRs.

```
exp <- count_matrix(exp, data_type = "tsr")
exp <- tmm_normalize(exp, data_type = "tsr")
```

### Correlation Matrix Plots

After TMM normalization, correlation plots can be generated.

```
p <- plot_correlation(exp, data_type = "tsr", font_size = 2, pt_size = 0.4) +
        ggplot2::theme_bw() +
        ggplot2::theme(text = element_text(size = 3), panel.grid = element_blank())

ggsave("tsr_correlation.png", plot = p, device = "png", type = "cairo", height = 2, width = 2)
```

![tsr_corr_plot](../inst/images/tsr_correlation.png)

## TSR Genomic Distribution

It is also expected for TSRs to be enriched upstream of annotated start codons.
TSSexploreR can generate many of the same plots used to explore TSS distribution for TSRs.

### Genomic Distribution Plot

A stacked bar plot can be generated to showcase the fractional distribution of TSRs relative to features.

```
tsr_distribution <- genomic_distribution(exp, data_type = "tsr", threshold = 3)

p <- plot_genomic_distribution(tsr_distribution) +
        ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tsr_genomic_distribution.png", plot = p, device = "png", type = "cairo", height = 1, width = 2.5)
```

![tsr_genomic_distribution](../inst/images/tsr_genomic_distribution.png)

### Feature Detection Plot

The number of genes, and fraction of genes with a promoter proximal TSRs can be made into a stacked bar plot.

```
features <- detect_features(exp, data_type = "tsr")

p <- plot_detected_features(features) +
        ggplot2::theme(text = element_text(size = 3), legend.key.size = unit(0.3, "cm"))

ggsave("tsr_feature_plot.png", plot = p, device = "png", type = "cairo", height = 1, width = 1.75)
```

![tsr_feature_plot](../inst/images/tsr_feature_plot.png)

### Average Plots

Average plots are a convenient way to summarize the average distribution of TSRs relative to start codons
or annotated TSSs.

```
p <- plot_average(exp, data_type = "tsr", ncol = 3) +
        ggplot2::theme(text = element_text(size = 4))

ggsave("tsr_average_plot.png", plot = p, device = "png", type = "cairo", height = 1, width = 2)
```

![tsr_average_plot](../inst/images/tsr_average_plot.png)

### Heatmaps

While an average plot may give a general overview of TSRs relative to annotated start codons or TSSs,
it may sometimes be appropriate to generate a heatmap with TSR positions for all features displayed.

```
count_matrix <- tsr_heatmap_matrix(exp, threshold = 3, upstream = 500, downstream = 500)

p <- plot_heatmap(count_matrix, ncol = 3, background_color = "white") +
        ggplot2::theme(text = element_text(size = 4), legend.key.size = unit(0.3, "cm"))

ggsave("tsr_heatmap.png", plot = p, device = "png", type = "cairo", height = 2, width = 4)
```

![tsr_heatmap](../inst/images/tsr_heatmap.png)

## TSR Metrics and Shape

After clustering the TSSs into TSRs, various descriptive and quantative measures were calculated.
TSSexploreR allows for the deep exploration of TSSs and TSRs relative to these metrics.

### Summary Plots

Overall summary plots can be generated for most TSR metrics.
In this example violin plot the log2 + 1 transformed TSR score and width are plotted.

```
p <- plot_tsr_metric(exp, tsr_metrics = c("score", "width"), log2_transform = TRUE, ncol = 2) +
	ggplot2::theme(text = element_text(size = 5), legend.key.size = unit(0.3, "cm"))

ggsave("tsr_metrics.png", plot = p, device = "png", type = "cairo", height = 1, width = 3)
```

![tsr_metrics](../inst/images/tsr_metrics.png)

### Descriptive Plots

Most plots in this vignette can be filtered, ordered, grouped, and quantiled
by any of the desired metrics.
Here is an example of what can be done with the TSS sequence motif plot using TSR metrics.
In this plot the dominant TSSs of each TSRs are only considered, and TSSs with a score below 10 are discarded.
The plot is then split by the shape class of the TSR.
A more detailed guide on advanced plotting can be found here.

```
assembly <- system.file("extdata", "S288C_Assembly.fasta", package = "tsrexplorer")

conditions <- list(order_by = "score", grouping = "shape_class")
seqs <- tss_sequences(
	exp, genome_assembly = assembly, threshold = 10,
	dominant = TRUE, data_conditions = conditions
)

p <- plot_sequence_logo(seqs, ncol = 3)

png("tss_seq_logo_conditioned.png", units = "in", res = 300, height = 2, width = 6, type = "cairo")
p
dev.off()
```

![tss_seq_logo_conditioned](../inst/images/tss_seq_logo_conditioned.png)

## Gene Tracks

Vieweing TSSs and/or TSRs in gene tracks can be useful for a variety of reason.
It can make obvious which 5' isoforms of a transcript are expressed,
hint at potential misannotation of genes, uncover 5' UTR structure, and other various goodies.

A convenient function to generate gene tracks for TSSs and/or TSRs is included in tsrexplorer.
Tracks can be created based on a gene/transcript name, or genomic coordinates.
Additionally, if tracks are generated based on genes/transcripts, the promoter region only can be optionally displayed.

```
annotation <- system.file("extdata", "S288C_Annotation.gtf", package = "tsrexplorer")

png("gene_track.png", units = "in", res = 300, height = 3.5, width = 4, type = "cairo")
gene_tracks(
	experiment, annotation, feature_name = "YAL012W",
	samples = c(
		"TSS:S288C_WT_1", "TSR:S288C_WT_1",
		"TSS:S288C_WT_2", "TSR:S288C_WT_2",
		"TSS:S288C_WT_3", "TSR:S288C_WT_3"
	),
	ymax = 80, tss_colors = viridis::viridis(3), tsr_colors = viridis::viridis(3)
)
dev.off()
```

![gene_track](../inst/images/gene_track.png)

